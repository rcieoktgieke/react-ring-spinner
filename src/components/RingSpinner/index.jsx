import React from 'react'

class RingSpinner extends React.Component {
	render () {
		const color1 = this.props.color1 ?
			this.props.color1
		 	: 'white'
		const color2 = this.props.color2 ?
			this.props.color2
		 	: 'white'
		const backgroundColor = this.props.backgroundColor ?
			this.props.backgroundColor
			: 'rgba(0, 0, 0, 0.9)'
		const strokeWidth = this.props.strokeWidth ?
		 	this.props.strokeWidth
			: 3
		const opacity1 = this.props.opacity1 ?
		 	this.props.opacity1
			: 0.9
		const opacity2 = this.props.opacity2 ?
		 	this.props.opacity2
			: 0.3
		const duration = this.props.duration ?
		 	this.props.duration
			: '1s'
		const opacityStartAdjust = this.props.opacityStartAdjust ?
		 	this.props.opacityStartAdjust
			: '0.35s'
		const circleRadius1 = this.props.circleRadius1 ?
		 	this.props.circleRadius1
			: 30
		const circleRadius2 = this.props.circleRadius2 ?
		 	this.props.circleRadius2
			: 45
		const keySplinesOut = this.props.keySplinesOut ?
			this.props.keySplinesOut
			: '0.8 0 0.2 1'
		const keySplinesIn = this.props.keySplinesIn ?
			this.props.keySplinesIn
			: '0.6 0 0.2 1'

		const spinnerID = Math.floor(Math.random() * 1000)
		return (
			<svg
				version='1.1'
				style={{
					width: '100%',
					height: '100%',
					backgroundColor: backgroundColor
				}}
				viewBox='0 0 100 100'
			>

				<circle
					cx='50'
					cy='50'
					r={circleRadius1}
					stroke={color1}
					opacity={opacity1}
					strokeWidth={strokeWidth}
					fill='none'
				>
					<animate
						attributeName='r'
						from={circleRadius1}
						to={circleRadius1}
						dur={duration}
						calcMode='spline'
						values={`${circleRadius1}; ${circleRadius2}; ${circleRadius1};`}
						keyTimes='0; 0.5; 1'
						keySplines={`${keySplinesOut};
												${keySplinesIn}`}
						repeatCount='indefinite'
					/>
					<animate
						attributeType='CSS'
						attributeName='opacity'
						from={opacity1}
						to={opacity1}
						dur={duration}
						calcMode='spline'
						values={`${opacity1}; ${opacity2}; ${opacity1};`}
						keyTimes='0; 0.5; 1'
						keySplines={`${keySplinesOut};
												${keySplinesIn}`}
						begin={opacityStartAdjust}
						repeatCount='indefinite'
					/>
				</circle>
				<circle
					cx='50'
					cy='50'
					r={circleRadius2}
					stroke={color2}
					opacity={opacity2}
					strokeWidth={strokeWidth}
					fill='none'
				>
					<animate
						attributeName='r'
						from={circleRadius2}
						to={circleRadius2}
						dur={duration}
						calcMode='spline'
						values={`${circleRadius2}; ${circleRadius1}; ${circleRadius2};`}
						keyTimes='0; 0.5; 1'
						keySplines={`${keySplinesIn};
												${keySplinesOut}`}
						repeatCount='indefinite'
					/>
					<animate
						attributeType='CSS'
						attributeName='opacity'
						from={opacity2}
						to={opacity2}
						dur={duration}
						calcMode='spline'
						values={`${opacity2}; ${opacity1}; ${opacity2};`}
						keyTimes='0; 0.5; 1'
						keySplines={`${keySplinesIn};
												${keySplinesOut}`}
						begin={opacityStartAdjust}
						repeatCount='indefinite'
					/>
				</circle>
			</svg>
		)
	}
}

export default RingSpinner
