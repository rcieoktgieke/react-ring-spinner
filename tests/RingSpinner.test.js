import React from 'react'
import RingSpinner from '../src/components/RingSpinner/index.jsx'
import renderer from 'react-test-renderer'

test('renders with default config', () => {
  const spinner = renderer
    .create(<RingSpinner />)
    .toJSON()
  expect(spinner).toMatchSnapshot()
})

test('renders with explicit default config', () => {
  const spinner = renderer
    .create(<RingSpinner
      backgroundColor={'rgba(0, 0, 0, 0.9)'}
      duration={'1s'}
      keySplinesIn={'0.6 0 0.2 1'}
      keySplinesOut={'0.8 0 0.2 1'}
      opacityAnimationDelay={'0.35s'}
      ring1Color={'white'}
      ring2Color={'white'}
      ringsOpacityMax={0.9}
      ringsOpacityMin={0.3}
      ringsRadiusMax={45}
      ringsRadiusMin={30}
      ringsStrokeWidth={'3px'}
    />)
    .toJSON()
  expect(spinner).toMatchSnapshot()
})

test('default config is unchanged', () => {
  const defaultSpinner = renderer
    .create(<RingSpinner />)
    .toJSON()
  const explicitSpinner = renderer
    .create(<RingSpinner
      backgroundColor={'rgba(0, 0, 0, 0.9)'}
      duration={'1s'}
      keySplinesIn={'0.6 0 0.2 1'}
      keySplinesOut={'0.8 0 0.2 1'}
      opacityAnimationDelay={'0.35s'}
      ring1Color={'white'}
      ring2Color={'white'}
      ringsOpacityMax={0.9}
      ringsOpacityMin={0.3}
      ringsRadiusMax={45}
      ringsRadiusMin={30}
      ringsStrokeWidth={'3px'}
    />)
    .toJSON()
  expect(defaultSpinner).toEqual(explicitSpinner)
})

test('renders with tweaked config values', () => {
  const spinner = renderer
    .create(<RingSpinner
      backgroundColor={'green'}
      duration={'2s'}
      keySplinesIn={'0.4 0.1 0.1 0.9'}
      keySplinesOut={'0.9 0.1 0.3 0.9'}
      opacityAnimationDelay={'0.3s'}
      ring1Color={'red'}
      ring2Color={'rgba(255, 0, 0, 0.9)'}
      ringsOpacityMax={1}
      ringsOpacityMin={0.7}
      ringsRadiusMax={50}
      ringsRadiusMin={40}
      ringsStrokeWidth={'1px'}
    />)
    .toJSON()
  expect(spinner).toMatchSnapshot()
})
