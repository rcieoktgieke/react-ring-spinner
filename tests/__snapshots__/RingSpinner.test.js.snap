// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with default config 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "rgba(0, 0, 0, 0.9)",
      "height": "100%",
      "width": "100%",
    }
  }
  version="1.1"
  viewBox="0 0 100 100"
>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={0.9}
    r={30}
    ringsStrokeWidth="3px"
    stroke="white"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="1s"
      from={30}
      keySplines="0.8 0 0.2 1;
												0.6 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={30}
      values="30; 45; 30;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.35s"
      calcMode="spline"
      dur="1s"
      from={0.9}
      keySplines="0.8 0 0.2 1;
												0.6 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={0.9}
      values="0.9; 0.3; 0.9;"
    />
  </circle>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={0.3}
    r={45}
    ringsStrokeWidth="3px"
    stroke="white"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="1s"
      from={45}
      keySplines="0.6 0 0.2 1;
												0.8 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={45}
      values="45; 30; 45;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.35s"
      calcMode="spline"
      dur="1s"
      from={0.3}
      keySplines="0.6 0 0.2 1;
												0.8 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={0.3}
      values="0.3; 0.9; 0.3;"
    />
  </circle>
</svg>
`;

exports[`renders with explicit default config 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "rgba(0, 0, 0, 0.9)",
      "height": "100%",
      "width": "100%",
    }
  }
  version="1.1"
  viewBox="0 0 100 100"
>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={0.9}
    r={30}
    ringsStrokeWidth="3px"
    stroke="white"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="1s"
      from={30}
      keySplines="0.8 0 0.2 1;
												0.6 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={30}
      values="30; 45; 30;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.35s"
      calcMode="spline"
      dur="1s"
      from={0.9}
      keySplines="0.8 0 0.2 1;
												0.6 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={0.9}
      values="0.9; 0.3; 0.9;"
    />
  </circle>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={0.3}
    r={45}
    ringsStrokeWidth="3px"
    stroke="white"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="1s"
      from={45}
      keySplines="0.6 0 0.2 1;
												0.8 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={45}
      values="45; 30; 45;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.35s"
      calcMode="spline"
      dur="1s"
      from={0.3}
      keySplines="0.6 0 0.2 1;
												0.8 0 0.2 1"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={0.3}
      values="0.3; 0.9; 0.3;"
    />
  </circle>
</svg>
`;

exports[`renders with tweaked config values 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "green",
      "height": "100%",
      "width": "100%",
    }
  }
  version="1.1"
  viewBox="0 0 100 100"
>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={1}
    r={30}
    ringsStrokeWidth="1px"
    stroke="red"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="2s"
      from={30}
      keySplines="0.9 0.1 0.3 0.9;
												0.4 0.1 0.1 0.9"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={30}
      values="30; 45; 30;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.3s"
      calcMode="spline"
      dur="2s"
      from={1}
      keySplines="0.9 0.1 0.3 0.9;
												0.4 0.1 0.1 0.9"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={1}
      values="1; 0.7; 1;"
    />
  </circle>
  <circle
    cx="50"
    cy="50"
    fill="none"
    opacity={0.7}
    r={45}
    ringsStrokeWidth="1px"
    stroke="rgba(255, 0, 0, 0.9)"
  >
    <animate
      attributeName="r"
      calcMode="spline"
      dur="2s"
      from={45}
      keySplines="0.4 0.1 0.1 0.9;
												0.9 0.1 0.3 0.9"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={45}
      values="45; 30; 45;"
    />
    <animate
      attributeName="opacity"
      attributeType="CSS"
      begin="0.3s"
      calcMode="spline"
      dur="2s"
      from={0.7}
      keySplines="0.4 0.1 0.1 0.9;
												0.9 0.1 0.3 0.9"
      keyTimes="0; 0.5; 1"
      repeatCount="indefinite"
      to={0.7}
      values="0.7; 1; 0.7;"
    />
  </circle>
</svg>
`;
